### PR TITLE
[#675] Add Templates Endpoints

### LINKED ISSUE #675: Add Templates Endpoints
### Templates

We must provide endpoints for:
- creating templates
- listing all templates
- listing one template
- updating templates
- deleting templates

In other words, we need the ability to store templates and the ability to retrieve templates. These will all be basic operations to manage data of a specific kind into the system.

It is important to state that we will not manage them, but rather the CMS will be external.

There should be no limit to how many templates can be saved or retrieved.

We need to offer the option to filter saved replies when they are being listed. This filtering system should work as either:
- filter by name
- filter by parts of a name

We need to offer the ability to store JSON that is source specific and then send it to conversations, if possible.

### PR DESCRIPTION
**Pull request by paulodiniz**
Fixes #675

### CODE REVIEW

**Comment by paulodiniz** [comment]
Points to discuss
-`/templates.list` is not paginated
-`/templates.list` filters by name in memory, and not by a lucene query as we wrote on the docs PR

If we decide to go with these 2 approaches, we also need to update the docs PR https://github.com/airyhq/airy/pull/872

File: backend/avro/communication/template.avsc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "namespace": "co.airy.avro.communication",
+  "name": "Template",
+  "type": "record",
+  "fields": [
+    {"name": "id", "type": "string"},
+    {"name": "name", "type": "string"},
+    {"name": "content", "type": "string"},
+    {"name": "variables", "type": "string"}
```
Comment (paulodiniz) [inline backend/avro/communication/template.avsc:9]: I'm thinking if there's a better way to store `variables`

File: backend/model/template/src/main/java/co/airy/model/template/TemplatePayload.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class TemplatePayload {
+    private String id;
+    private String name;
+    private String content;
+    private JsonNode variables;
+    public static TemplatePayload fromTemplate(final Template template) {
+        final ObjectMapper mapper = new ObjectMapper();
```
Comment (paulodiniz) [inline backend/model/template/src/main/java/co/airy/model/template/TemplatePayload.java:23]: Maybe we can use the objectmapper we. already have

Comment (chrismatix) [reply backend/model/template/src/main/java/co/airy/model/template/TemplatePayload.java:23]: Hmm, I wouldn't. The templates are source specific message content. So we want to pass them as unmodified as possible, right?

File: backend/api/admin/src/main/java/co/airy/core/api/admin/TemplatesController.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            return ResponseEntity.notFound().build();
+        }
+        stores.deleteTemplate(template);
+        return ResponseEntity.ok(new EmptyResponsePayload());
+    }
+
+    @PostMapping("/templates.list")
+    ResponseEntity<?> listTemplates(@RequestBody @Valid ListTemplatesRequestPayload payload) {
+        final List<TemplatePayload> response = stores.getTemplates().stream()
+                .filter(t -> t.getName().contains(payload.getName()))
```
Comment (chrismatix) [inline backend/api/admin/src/main/java/co/airy/core/api/admin/TemplatesController.java:78]: Will this work, if the user does not pass the filter? 🤔

**Review by chrismatix — APPROVED** [review]
LGTM

File: backend/model/template/src/main/java/co/airy/model/template/TemplatePayload.java (hunk 2 above)
Comment (paulodiniz) [reply backend/model/template/src/main/java/co/airy/model/template/TemplatePayload.java:23]: Good point 👍

**Review by chrismatix — APPROVED** [review]

**Pull request merged**